packages/astro/src/core/dev/index.ts

```diff
@@ -329,16 +329,25 @@ export class AstroDevServer {
       res.write(html);
       res.end();
     } catch (err: any) {
-      this.viteServer.ws.send({ type: 'error', err });
       const statusCode = 500;
-      const html = serverErrorTemplate({
+      const mod = filePath && this.viteServer.moduleGraph.getModuleById(viteifyURL(filePath));
+      if (mod) {
+        for (const m of [mod, ...mod.importedModules]) {
+          this.viteServer.moduleGraph.invalidateModule(m);
```

Comment:
This fixes the `Expected an exported Astro component but received typeof undefined` error